fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexFunctionHelper.java

```diff
@@ -154,34 +159,72 @@ public static KeyExpression getGroupedKey(@Nonnull KeyExpression key) {
-            if (start == 0 && end == 1) {
-                return key;
-            }
-            throw new RecordCoreException("grouping breaks apart key other than Then");
+        if (start == 0 && end == key.getColumnSize()) {
+            return key;
+        }
+        if (key instanceof ThenKeyExpression) {
+            return getThenSubKey((ThenKeyExpression)key, start, end);
+        }
+        if (key instanceof NestingKeyExpression) {
+            final NestingKeyExpression nesting = (NestingKeyExpression)key;
+            return new NestingKeyExpression(nesting.getParent(), getSubKey(nesting.getChild(), start, end));
         }
-        ThenKeyExpression then = (ThenKeyExpression)key;
-        List<KeyExpression> children = then.getChildren();
-        if (end == start + 1) {
-            return children.get(start);
+        throw new RecordCoreException("grouping breaks apart key other than Then");
+    }
+
+    protected static KeyExpression getThenSubKey(ThenKeyExpression then, int columnStart, int columnEnd) {
+        final List<KeyExpression> children = then.getChildren();
+        int columnPosition = 0;
+        int startChildPosition = -1;
+        int startChildStart = -1;
+        int startChildEnd = -1;
+        for (int childPosition = 0; childPosition < children.size(); childPosition++) {
+            final KeyExpression child = children.get(childPosition);
+            final int childColumns = child.getColumnSize();
+            if (startChildPosition < 0 && columnPosition + childColumns > columnStart) {
+                startChildPosition = childPosition;
+                startChildStart = columnStart - columnPosition;
+                startChildEnd = childColumns;
+            }
+            if (columnPosition + childColumns >= columnEnd) {
+                int endChildEnd = columnEnd - columnPosition;
+                if (childPosition == startChildPosition) {
+                    // Just one child spans column start, end.
+                    if (startChildPosition == 0 && endChildEnd == childColumns) {
+                        return child;
+                    } else {
+                        return getSubKey(child, startChildStart, endChildEnd);
+                    }
+                }
+                if (startChildStart == 0 && endChildEnd == childColumns) {
+                    return new ThenKeyExpression(children.subList(startChildPosition, childPosition + 1));
+                }
+                final List<KeyExpression> keys = new ArrayList<>(childPosition - startChildPosition + 1);
+                keys.add(getSubKey(children.get(startChildPosition), startChildStart, startChildEnd));
+                if (childPosition > startChildPosition + 1) {
+                    keys.addAll(children.subList(startChildPosition + 1, childPosition));
+                }
+                keys.add(getSubKey(child, 0, endChildEnd));
+                return new ThenKeyExpression(keys);
```

Comment:
This looks right to me, but I think it could use more coverage on the various cases to validate that